zkevm-circuits/src/evm_circuit/execution.rs

```diff
@@ -0,0 +1,727 @@
+            signextend_gadget: configure_gadget!(),
+            stop_gadget: configure_gadget!(),
+            swap_gadget: configure_gadget!(),
+            step: step_curr,
+            presets_map,
+        };
+
+        Self::configure_lookup(
+            meta,
+            q_step,
+            fixed_table,
+            tx_table,
+            rw_table,
+            bytecode_table,
+            independent_lookups,
+        );
+
+        config
+    }
+
+    fn configure_gadget<G: ExecutionGadget<F>>(
+        meta: &mut ConstraintSystem<F>,
+        q_step: Selector,
+        randomness: &Expression<F>,
+        step_curr: &Step<F>,
+        step_next: &Step<F>,
+        independent_lookups: &mut Vec<Vec<Lookup<F>>>,
+        presets_map: &mut HashMap<ExecutionState, Vec<Preset<F>>>,
+    ) -> G {
+        let mut cb = ConstraintBuilder::new(
+            step_curr,
+            step_next,
+            randomness.clone(),
+            G::EXECUTION_STATE,
+        );
+
+        let gadget = G::configure(&mut cb);
+
+        let (constraints, lookups, presets) = cb.build();
+        assert!(
+            presets_map.insert(G::EXECUTION_STATE, presets).is_none(),
+            "execution result already configured"
+        );
+
+        if !constraints.is_empty() {
+            meta.create_gate(G::NAME, |meta| {
+                let q_step = meta.query_selector(q_step);
+
+                constraints.into_iter().map(move |(name, constraint)| {
+                    (name, q_step.clone() * constraint)
+                })
+            });
+        }
+
+        independent_lookups.push(lookups);
+
+        gadget
+    }
+
+    fn configure_lookup<TxTable, RwTable, BytecodeTable>(
+        meta: &mut ConstraintSystem<F>,
+        q_step: Selector,
+        fixed_table: [Column<Fixed>; 4],
+        tx_table: TxTable,
+        rw_table: RwTable,
+        bytecode_table: BytecodeTable,
+        independent_lookups: Vec<Vec<Lookup<F>>>,
+    ) where
+        TxTable: LookupTable<F, 4>,
+        RwTable: LookupTable<F, 8>,
+        BytecodeTable: LookupTable<F, 4>,
+    {
+        let mut input_exprs_map = HashMap::new();
+
+        for lookups in independent_lookups {
+            let mut index_map = HashMap::new();
+
+            for lookup in lookups {
+                let table = lookup.table();
+                let input_exprs =
+                    input_exprs_map.entry(table).or_insert_with(Vec::new);
+                let index = index_map.entry(table).or_insert(0);
+
+                if *index == input_exprs.len() {
+                    input_exprs.push(lookup.input_exprs());
+                } else {
+                    for (acc, expr) in input_exprs[*index]
+                        .iter_mut()
+                        .zip(lookup.input_exprs().into_iter())
+                    {
+                        *acc = acc.clone() + expr;
+                    }
+                }
+                *index += 1;
+            }
+        }
```

Comment:
This took me a while to understand what exactly this does and why it works. Perhaps some comments for posterity pointing in the right direction could help speed that up. :) Something like (if I even understand correctly): lookups are reused between execution states when possible, and lookup expressions from different execution states can be added together because they are conditioned on their execution state.

Reply:
Sorry for the spaghetti code here, will doc this now.

You are totally right, the lookup here is try to add up all **independent lookup** input expression together, to use as less lookup argument as possible (but still we are abusing lookup). And the reason they are independent is because we have constraint one and only one of execution state would be enabled at a step.

Reply:
But there could be the case that none of the **independent lookup** is enabled for this table at a step, right? So does it mean that we should add a tuple of all zeros to each lookup table?

Reply:
I think there are 2 questions mixed here:

> But there could be the case that none of the independent lookup is enabled for this table at a step

Yeah, in this case, the lookup argument will then not configured.

> does it mean that we should add a tuple of all zeros to each lookup table?

This is already done [here](https://github.com/appliedzkp/zkevm-circuits/pull/196/files#diff-b4e5f73a1df679165a0103a62054eeff42c1bcbba32eff0ba8287c52740f2738R65-R70), [here](https://github.com/appliedzkp/zkevm-circuits/pull/196/files#diff-b4e5f73a1df679165a0103a62054eeff42c1bcbba32eff0ba8287c52740f2738R158-R165), [here](https://github.com/appliedzkp/zkevm-circuits/pull/196/files#diff-b4e5f73a1df679165a0103a62054eeff42c1bcbba32eff0ba8287c52740f2738R197-R204) and [here](https://github.com/appliedzkp/zkevm-circuits/pull/196/files#diff-b4e5f73a1df679165a0103a62054eeff42c1bcbba32eff0ba8287c52740f2738R237-R244).